smarts/env/wrappers/parallel_env.py

```diff
@@ -0,0 +1,355 @@
+                f"CPU cores ({mp.cpu_count()}).",
+                ResourceWarning,
+            )
+
+        if any([not callable(ctor) for ctor in env_constructors]):
+            raise TypeError(
+                f"Found non-callable `env_constructors`. Expected `env_constructors` of type "
+                f"`Sequence[Callable[[], gym.Env]]`, but got {env_constructors})."
+            )
+
+        # Worker polling period in seconds.
+        self._polling_period = 0.1
+
+        # Fork is not a thread safe method.
+        forkserver_available = "forkserver" in mp.get_all_start_methods()
+        start_method = "forkserver" if forkserver_available else "spawn"
+        mp_ctx = mp.get_context(start_method)
+
+        self.env_constructors = env_constructors
+
+        self.error_queue = mp_ctx.Queue()
+        self.parent_pipes = []
+        self.processes = []
+        for idx, env_constructor in enumerate(self.env_constructors):
+            parent_pipe, child_pipe = mp_ctx.Pipe()
+            process = mp_ctx.Process(
+                target=_worker,
+                name=f"Worker<{type(self).__name__}>-<{idx}>",
+                args=(
+                    idx,
+                    sim_name,
+                    CloudpickleWrapper(env_constructor),
+                    auto_reset,
+                    child_pipe,
+                    self.error_queue,
+                    self._polling_period,
+                ),
+            )
+            self.parent_pipes.append(parent_pipe)
+            self.processes.append(process)
+
+            # Daemonic subprocesses quit when parent process quits. However, daemonic
+            # processes cannot spawn children. Hence, `process.daemon` is set to False.
+            process.daemon = False
+            process.start()
+            child_pipe.close()
+
+        self._state = AsyncState.DEFAULT
+
+        # Wait for all environments to successfully startup
+        _, successes = zip(*[pipe.recv() for pipe in self.parent_pipes])
+        self._raise_if_errors(successes)
+
+        # Get and check observation and action spaces
+        observation_space, action_space = self._get_spaces()
+
+        # TODO: This dummy `observation_space` and `action_space` should be removed after they
+        # are properly specified in SMARTS/smarts/env/hiway_env.py:__init__() function.
+        action_space = gym.spaces.Dict(
+            {
+                agent_id: gym.spaces.Box(
+                    np.array([0, 0, -1]), np.array([+1, +1, +1]), dtype=np.float32
+                )  # throttle, break, steering
+                for agent_id in ["Agent1", "Agent2"]
+            }
+        )
+        observation_space = gym.spaces.Box(
+            low=-1, high=1, shape=(256, 256, 3), dtype=np.float32
+        )
+
+        super(AsyncVectorEnv, self).__init__(
+            num_envs=len(env_constructors),
+            observation_space=observation_space,
+            action_space=action_space,
+        )
```

Comment:
This seems dangerous -- that is, to "skip over" calling `super(ParralelEnv, self).__init__` and instead call `super(AsyncVectorEnv,  self).__init__`.  I.e., to call the init function from `VectorEnv` instead of calling the one from the class's parent, `AsyncVectorEnv`.   At a minimum, there should be a comment here as to why this was done (and warning future coders of the anomaly, which they might overlook).

Is it to avoid having to support members of the base class such as `copy` and `shared_memory`?

The danger here is that this is a way of "breaking encapsulation".  You have to look at the source code of the parent class(es) to know whether what we've done here is ok (it seems to be right now), and it may not continue to be so if either of them change.   I see that we've pinned to this version of `gym` (`0.18.3`) so we are probably able to get away with this for now, but code like this increases the burden of doing any `gym` upgrade in the future because the person doing that may not understand the subtleties of what you've done here.   Also, by skipping its `__init__`, we are inheriting an "incomplete" version of the base class (since its `__init__` method does set fields that we don't), which may confuse any code that wants to reference our `ParallelEnv` as a `AsyncVectorEnv`.

Reply:
Obsolete.